packages/sitecore-jss/src/personalization/layout-personalization-utils.ts

```diff
@@ -0,0 +1,109 @@
+import {
+  PersonalizedComponentRendering,
+  isComponentRendering,
+  isPersonalizedComponentRendering,
+  ComponentRendering,
+  PlaceholdersData,
+  hasPersonalization
+} from '../layout/models';
+
+export class LayoutPersonalizationUtils {
+
+  buildPersonalizedFragment(uid: string, personalizedFragments: { [key: string]: ComponentRendering | null | undefined; }, defaultComponent: ComponentRendering | null): ComponentRendering | null {
+    var personalizedFragment = personalizedFragments[uid];
+    if (personalizedFragment === null) {
+      return null;
+    } else if (personalizedFragment === undefined) {
+      return defaultComponent;
+    }
+
+    function replacePersonalizedRenderings(context: ComponentRendering): boolean {
+      if (context.placeholders) {
+        for (const key in context.placeholders) {
+          if (context.placeholders.hasOwnProperty(key)) {
+            const placeholder = context.placeholders[key];
+            var hiddenComponents: any[] = [];
+            for (let i = 0; i < placeholder.length; i++) {
+              var component = placeholder[i];
+              if (requirePersonalization(component)) {
+                var personalizedFragment = personalizedFragments[component.uid];
+                if (personalizedFragment === null) {
+                  hiddenComponents.push(component);
+                } else if (personalizedFragment) {
+                  var wasReplaced = replacePersonalizedRenderings(personalizedFragment);
+                  if (!wasReplaced) {
+                    return false;
+                  }
+                  placeholder[i] = personalizedFragment;
+                } else {
+                  return false;
+                }
+              }
+            }
+            if (hiddenComponents.length) {
+              context.placeholders[key] = placeholder.filter(function (item: any) {
+                return hiddenComponents.indexOf(item) == -1;
+              });
+            }
+          }
+        }
+      }
+      return true;
```

Comment:
The `replacePersonalizedRenderings` function body is difficult to read. Add line spacing. Reduce nesting.